docs/design/2018-07-22-enhance-propagations.md

```diff
@@ -0,0 +1,189 @@
+# Proposal: Enhance constraint propagation in TiDB logical plan
+
+- Author(s):     [@bb7133](https://github.com/bb7133), [@zz-jason](https://github.com/zz-jason) 
+- Last updated:  2018-07-22
+- Discussion at: https://docs.google.com/document/d/1G3wVBaiza9GI5q9nwHLCB2I7r1HUbg6RpTNFpQWTFhQ/edit#
+
+## Abstract
+
+This proposal tries to illustrate some rules that can be added to current constraint/filter propgation optimizations in TiDB logcial plan.
+
+## Background
+
+Currently, most of the constraint propagation work in TiDB is done by `propagateConstantSolver`, which does:
+
+1. Find the `column = constant` expression and substitute the constant for the column, as well as try to fold the substituted constant expression if possible, for example:
+
+ Given `a = b and a = 2 and b = 3`, it becomes `2 = 3` after substitution and leads to a final `false` constant.
+
+2. Find the `column A = column B` expression (which happens in `join` statements mostly) and propagate expressions like `column op constant` (as well as `constant op column`) based on the equality relation. The supported operators are:
+
+ * `ast.LT('<')`
+ * `ast.GT('>')`
+ * `ast.LE('<=')`
+ * `ast.GE('>=')`
+ * `ast.NE('!=')`
+
+The `propagateConstantSolver` makes more detailed/explicit filters/constraints, which can be used within other optimization rules. For example, in the predicate-pushdown optimization, it generates more predicates that can be pushed closer to the data source (TiKV), and thus reduces the amount of data in the whole data path.
+
+We can further do the optimization by introducing more rules and inferring/propagating more constraints from the existing ones, which helps us build a better logical plan.
+
+## Proposal
+
+Here are proposed rules we can consider:
+
+1. Infer more filters/constraints from column equality relation
```

Comment:
We need to be EXTREMELY careful here. Consider the following examples:
```
t1.a = t2.a and t1.a is null
```
we should NOT infer `t2.a is null`, it is not logically equivalent because `null = null` is false;

another example is:
```
t1.a = t2.a and cast(t1.a as char(10)) = '+0.0'
```
we should NOT infer `cast(t2.a as char(10)) = '+0.0'`, it is not logically equivalent either because this would filter out tuples with `t2.a = -0.0`

These 2 cases should fail in current master branch logically, we need to fix them.

Maybe there are other failed cases as well, once more, we need to think over this carefully.

Reply:
I see, you're right! I tried to get some failed corner cases but didn't get the ones you commented. Thank you~

Reply:
I ported this proposal back to original google doc: https://docs.google.com/document/d/1G3wVBaiza9GI5q9nwHLCB2I7r1HUbg6RpTNFpQWTFhQ/edit?usp=sharing, and made updates based on your cases. As you said, we should think very carefully about this part, please let me know if you have any new idea, thanks